File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";

        int[] hash = new int[128]; // ASCII size
        for (char c : t.toCharArray()) {
            hash[c]++;
        }

        int left = 0;
        int right = 0;
        int count = t.length(); // total chars to match
        int minLen = Integer.MAX_VALUE;
        int start = 0;

        while (right < s.length()) {
            char rChar = s.charAt(right);
            if (hash[rChar] > 0) count--;
            hash[rChar]--; // Always decrement, even if not in t
            right++;

            while (count == 0) {
                if (right - left < minLen) {
                    minLen = right - left;
                    start = left;
                }

                char lChar = s.charAt(left);
                hash[lChar]++;
                if (hash[lChar] > 0) count++; // Needed char removed
                left++;
            }
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start + minLen);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {

    let left = 0;
    let right  = 0;
    let count =0;
    let minLen = Infinity;
    let startIndex = -1;
    let required = t.length;
    let hash = new Array(128).fill(0);

    for(let i=0;i<t.length;i++){
        hash[t.charCodeAt(i)]++;
    }

    while(right < s.length){
        if(hash[s.charCodeAt(right)] > 0){
            count++;
        }
        hash[s.charCodeAt(right)]--;
        right++;

        while(count === required){
            if(right - left < minLen){
                minLen = right - left;
                startIndex = left;
            }
            hash[s.charCodeAt(left)]++;
            if(hash[s.charCodeAt(left)] > 0) count--;
            left++;
        }
    }

    return startIndex == -1 ? "" : s.substring(startIndex, startIndex + minLen);

    
};